src/evaluation/static_evaluation.h

```diff
@@ -22,45 +23,86 @@ namespace evaluation {
-    APPLY_SCORE(getChecksScore, board, ctx);
-    APPLY_SCORE(getPawnPushThreatScore, board, ctx);
-    APPLY_SCORE(getPassedPawnsScore, board, ctx);
-
-    const Score evaluation = score.phaseScore(phaseScore);
-    return board.player == PlayerWhite ? evaluation : -evaluation;
-}
+        uint8_t phaseScore = 0;
+        auto ctx = prepareContext(board);
+
+        /* terms that are not using ctx */
+        APPLY_SCORE(getTempoScore, board);
+
+        /* when tuning, we should never try to use the cache - we need to always update the tracer */
+#ifdef TUNING
+        APPLY_SCORE(getStaticKingPawnScore, board, ctx);
+#else
+        score += fetchOrStoreKpCache(board, ctx);
+#endif
+
+        /* piece scores - should be computed first as they populate ctx */
+        APPLY_SCORE(getKnightScore, board, ctx, phaseScore);
+        APPLY_SCORE(getBishopScore, board, ctx, phaseScore);
+        APPLY_SCORE(getRookScore, board, ctx, phaseScore);
+        APPLY_SCORE(getQueenScore, board, ctx, phaseScore);
+        APPLY_SCORE(getKingScore, board, ctx);
+
+        /* terms that consume ctx */
+        APPLY_SCORE(getKingZoneScore, ctx);
+        APPLY_SCORE(getPieceAttacksScore, board, ctx);
+        APPLY_SCORE(getChecksScore, board, ctx);
+        APPLY_SCORE(getPawnPushThreatScore, board, ctx);
+        APPLY_SCORE(getPassedPawnsScore, board, ctx);
+
+        const Score evaluation = score.phaseScore(phaseScore);
+        return board.player == PlayerWhite ? evaluation : -evaluation;
+    }
+
+private:
+    TermContext prepareContext(const BitBoard& board) const
+    {
+        auto ctx = TermContext {
+            .pawnAttacks { attackgen::getPawnAttacks<PlayerWhite>(board), attackgen::getPawnAttacks<PlayerBlack>(board) },
+            .kingZone { attackgen::getKingAttacks<PlayerWhite>(board), attackgen::getKingAttacks<PlayerBlack>(board) },
+            .attacksToKingZone { 0, 0 },
+            .pieceAttacks {},
+            .threats { 0, 0 },
+            .passedPawns { 0, 0 },
+        };
+
+        ctx.attacksToKingZone[PlayerWhite] += std::popcount(ctx.kingZone[PlayerWhite] & ctx.pawnAttacks[PlayerBlack]);
+        ctx.attacksToKingZone[PlayerBlack] += std::popcount(ctx.kingZone[PlayerBlack] & ctx.pawnAttacks[PlayerWhite]);
+
+        ctx.pieceAttacks[PlayerWhite][Pawn] = ctx.pawnAttacks[PlayerWhite];
+        ctx.pieceAttacks[PlayerBlack][Pawn] = ctx.pawnAttacks[PlayerBlack];
+
+        ctx.threats[PlayerWhite] |= ctx.pawnAttacks[PlayerWhite];
+        ctx.threats[PlayerBlack] |= ctx.pawnAttacks[PlayerBlack];
+
+        return ctx;
+    }
+
+    TermScore fetchOrStoreKpCache(const BitBoard& board, TermContext& ctx)
+    {
+        TermScore score(0, 0);
+
+        const auto kpProbe = m_kpCache.probe(board.kpHash);
+        if (kpProbe.has_value()) {
+            score += kpProbe->score;
+            ctx.passedPawns = kpProbe->passedPawns;
+        } else {
+            APPLY_SCORE(getStaticKingPawnScore, board, ctx);
```

Comment:
Could you please explain what's happening for a cache miss?

Reply:
We need to populate the cache before we can retrieve it.
So either we just bite the apple and do as before - or maybe we're able to fetch a cache. Best case it's faster, worst case it's same as before this PR :)

Reply:
I've added more information to the commit message as well.
But the theory is that the king pawn structure is more or less static. There's only so few possibilities (pawns can only move one step and only one direction).
So previously the pawn evaluation was more or less static - I've prepared for this already by adding more dynamic or volatile terms as non-piece terms: eg passed pawns where we need the full context to evaluate.

But for these "static terms" it's a nice and relatively simple way to get some speed ups. Also the cache is dum as can be - but it's still very efficient which just proves the point :)